Mathlib/Algebra/Polynomial/Bivariate.lean

```diff
@@ -0,0 +1,136 @@
+/-
+Copyright (c) 2024 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.RingTheory.AdjoinRoot
+
+/-!
+# Bivariate polynomials
+
+This file introduces the notation `R[X][Y]` for the polynomial ring `R[X][X]` in two variables,
+and the notation `Y` for the second variable, in the `Polynomial` scope.
+
+It also defines `Polynomial.evalEval` for the evaluation of a bivariate polynomial at a point
+on the affine plane, which is a ring homomorphism (`Polynomial.evalEvalRingHom`), as well as
+the abbreviation `CC` to view a constant in the base ring `R` as a bivariate polynomial.
+-/
+
+/-- The notation `Y` for `X` in the `Polynomial` scope. -/
+scoped[Polynomial] notation3:max "Y" => Polynomial.X (R := Polynomial _)
+
+/-- The notation `R[X][Y]` for `R[X][X]` in the `Polynomial` scope. -/
+scoped[Polynomial] notation3:max R "[X][Y]" => Polynomial (Polynomial R)
+
+namespace Polynomial
+
+noncomputable section
+
+variable {R S : Type*}
+
+section Semiring
+
+variable [Semiring R]
+
+/-- `evalEval x y p` is the evaluation `p(x,y)` of a two-variable polynomial `p : R[X][Y]`. -/
+abbrev evalEval (x y : R) (p : R[X][Y]) : R := eval x (eval (C y) p)
+
+/-- A constant viewed as a polynomial in two variables. -/
+abbrev CC (r : R) : R[X][Y] := C (C r)
+
+lemma evalEval_C (x y : R) (p : R[X]) : (C p).evalEval x y = p.eval x := by
+  rw [evalEval, eval_C]
+
+lemma evalEval_X (x y : R) : X.evalEval x y = y := by
+  rw [evalEval, eval_X, eval_C]
```

Comment:
We should add the behaviour of evaluations over the usual operations in a separate PR. Here are some to begin with (because it seems that these are the only ones I need):
```
lemma evalEval_add {R : Type u} [Semiring R] (x y : R) (p q : R[X][Y]) :
    (p + q).evalEval x y = p.evalEval x y + q.evalEval x y := by
  simp only [evalEval, eval_add]

lemma evalEval_sub {R : Type u} [Ring R] (x y : R) (p q : R[X][Y]) :
    (p - q).evalEval x y = p.evalEval x y - q.evalEval x y := by
  simp only [evalEval, eval_sub]

lemma evalEval_mul {R : Type u} [CommSemiring R] (x y : R) (p q : R[X][Y]) :
    (p * q).evalEval x y = p.evalEval x y * q.evalEval x y := by
  simp only [evalEval, eval_mul]

lemma evalEval_pow {R : Type u} [CommSemiring R] (x y : R) (p : R[X][Y]) (n : ℕ) :
    (p ^ n).evalEval x y = p.evalEval x y ^ n := by
  simp only [evalEval, eval_pow]
```

Reply:
You can also get these by rw [<- evalEvalRingHom] and then use the generic map_* lemmas ...

Reply:
If you do want them you can submit a PR and see how it goes.